tests/e2e/controller_test.go

```diff
@@ -228,27 +212,91 @@ func TestMain(m *testing.M) {
 	flag.BoolVar(&testOpts.operatorControllerTest, "test-operator-controller", true, "run operator controller tests")
 	flag.BoolVar(&testOpts.webhookTest, "test-webhook", true, "run webhook tests")
 
-	componentNames := strings.Join(maps.Keys(componentsTestSuites), ", ")
-	flag.Var(&testOpts.components, "test-component", "run tests for the specified component. valid components names are: "+componentNames)
+	componentNames := strings.Join(testOpts.components.Names(), ", ")
+	flag.BoolVar(&testOpts.components.enabled, "test-components", testOpts.components.enabled, "enable tests for components")
+	flag.Var(&testOpts.components.flags, "test-component", "run tests for the specified component. valid components names are: "+componentNames)
 
-	serviceNames := strings.Join(maps.Keys(servicesTestSuites), ", ")
-	flag.Var(&testOpts.services, "test-service", "run tests for the specified service. valid service names are: "+serviceNames)
+	serviceNames := strings.Join(testOpts.services.Names(), ", ")
+	flag.BoolVar(&testOpts.services.enabled, "test-services", testOpts.services.enabled, "enable tests for services")
+	flag.Var(&testOpts.services.flags, "test-service", "run tests for the specified service. valid service names are: "+serviceNames)
 
 	flag.Parse()
 
-	for _, n := range testOpts.components {
-		if _, ok := componentsTestSuites[n]; !ok {
-			fmt.Printf("test-component: unknown component %s, valid values are: %s", n, componentNames)
-			os.Exit(1)
+	if err := testOpts.components.Validate(); err != nil {
+		fmt.Printf("test-component: %s", err.Error())
+		os.Exit(1)
+	}
+
+	if err := testOpts.services.Validate(); err != nil {
+		fmt.Printf("test-service: %s", err.Error())
+		os.Exit(1)
+	}
+
+	os.Exit(m.Run())
+}
+
+type TestGroup struct {
+	name      string
+	enabled   bool
+	scenarios map[string]TestFn
+	flags     arrayFlags
+}
+
+func (tg *TestGroup) String() string {
+	return tg.name
+}
+
+func (tg *TestGroup) Names() []string {
+	return maps.Keys(tg.scenarios)
+}
+
+func (tg *TestGroup) Validate() error {
+	if tg.enabled == false && len(tg.flags) != 0 {
+		return errors.New("enabling individual scenarios is not supported when the entire group is disabled")
+	}
+
+	for _, n := range tg.flags {
+		n = strings.TrimLeft(n, "!")
+		if _, ok := tg.scenarios[n]; !ok {
```

Comment:
slices.Contains(tg.Names(), n) could be more readable?

Reply:
But that would be O(n) and some allocations vs O(1), which probably does not matter much in this case, no strong opinions

Reply:
> which probably does not matter much in this case

that were my thoughts

Reply:
ok, I'll change it